lib/src/chart/base/axis_chart/axis_chart_scaffold_widget.dart

```diff
@@ -18,105 +29,288 @@ import 'package:flutter/material.dart';
 ///
 /// `left`, `top`, `right`, `bottom` are some place holders to show titles
 /// provided by [AxisChartData.titlesData] around the chart
-/// `chart` is a centered place holder to show a raw chart.
-class AxisChartScaffoldWidget extends StatelessWidget {
+/// `chart` is a centered place holder to show a raw chart. The chart is
+/// built using [chartBuilder].
+class AxisChartScaffoldWidget extends StatefulWidget {
   const AxisChartScaffoldWidget({
     super.key,
-    required this.chart,
+    required this.chartBuilder,
     required this.data,
-  });
-  final Widget chart;
+    this.transformationController,
+    this.scaleAxis = ScaleAxis.none,
+    this.maxScale = 2.5,
+    this.minScale = 1,
+    this.trackpadScrollCausesScale = false,
+  })  : assert(minScale >= 1, 'minScale must be greater than or equal to 1'),
+        assert(
+          maxScale >= minScale,
+          'maxScale must be greater than or equal to minScale',
+        );
+
+  /// The builder to build the chart.
+  final ChartBuilder chartBuilder;
+
+  /// The data to build the chart.
   final AxisChartData data;
 
+  /// The transformation controller to control the transformation of the chart.
+  final TransformationController? transformationController;
+
+  /// Determines what axis should be scaled.
+  final ScaleAxis scaleAxis;
+
+  /// The maximum scale of the chart.
+  ///
+  /// Ignored when [scaleAxis] is [ScaleAxis.none].
+  final double maxScale;
+
+  /// The minimum scale of the chart.
+  ///
+  /// Ignored when [scaleAxis] is [ScaleAxis.none].
+  final double minScale;
+
+  /// Whether trackpad scroll causes scale.
+  ///
+  /// Ignored when [scaleAxis] is [ScaleAxis.none].
+  final bool trackpadScrollCausesScale;
+
+  @override
+  State<AxisChartScaffoldWidget> createState() =>
+      _AxisChartScaffoldWidgetState();
+}
+
+class _AxisChartScaffoldWidgetState extends State<AxisChartScaffoldWidget> {
+  late TransformationController _transformationController;
+
+  final _chartKey = GlobalKey();
+
+  Rect? _chartRect;
+
+  bool get _canScaleHorizontally =>
+      widget.scaleAxis == ScaleAxis.horizontal ||
+      widget.scaleAxis == ScaleAxis.free;
+
+  bool get _canScaleVertically =>
+      widget.scaleAxis == ScaleAxis.vertical ||
+      widget.scaleAxis == ScaleAxis.free;
+
+  @override
+  void initState() {
+    super.initState();
+    _transformationController =
+        widget.transformationController ?? TransformationController();
+    _transformationController.addListener(_updateChartRect);
+    updateRectPostFrame();
+  }
+
+  @override
+  void dispose() {
+    _transformationController.dispose();
```

Comment:
I think we should not dispose the controller without checking if it is the controller comes from the user, or we created it.

As I see, in TextField for example, we have a getter that returns the user one, and if user doesn't provide, it uses the default one (just like what we do)

https://github.com/flutter/flutter/blob/4fcf6333d2170818fe63521dbe9d84f3aeaa62ab/packages/flutter/lib/src/material/text_field.dart#L971-L972

And here it only disposes the one that is created internally.

https://github.com/flutter/flutter/blob/4fcf6333d2170818fe63521dbe9d84f3aeaa62ab/packages/flutter/lib/src/material/text_field.dart#L1179

So I think the responsibility for disposing a controller, is with the creator. (because they might re-use it I think)

Reply:
Nice!